### DISCUSSION TITLE
How to modify the headers of a response?

### DISCUSSION THREAD

**Discussion by FlysoftBeta** [tauri-apps/tauri#6187] [Q&A] [answered]
I need to set "Access-Control-Allow-Origin" for the response, but I can't find `add_WebResourceResponseReceived` in ICoreWebView2.

**Comment by FabianLars** [comment, marked as answer]
The `WebResourceResponseReceived` event can not be used to _modify_ the Responses, only to monitor them. If you want to modify a Response, you need to listen to the `WebResourceResourceRequested` event, cancel the request, do your own request stuff (via `reqwest` for example), modify that response and then send it back to the webview. Example from Microsoft docs: <https://learn.microsoft.com/en-us/microsoft-edge/webview2/how-to/webresourcerequested?tabs=dotnet#example-overriding-a-response-to-proactively-replace-it>

And the upstream feature request: <https://github.com/MicrosoftEdge/WebView2Feedback/issues/2355> - Note that this does _not_ exist on the linux and macos webviews either (but worse because you can't modify requests either).

Also in case it may be helpful for other issues, you need to use the `.cast` method of the windows-rs crate to cast the ICoreWebView2 to [ICoreWebView2_2](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2_2?view=webview2-1.0.1518.46#add_webresourceresponsereceived). Something like this

```rs
let core_2 = webview // `webview` is the arg you get in tauri's window.with_webview method.
  .controller()
  .CoreWebView2()
  .unwrap()
  .cast::<ICoreWebView2_2>()
  .unwrap();
```

**Reply by FlysoftBeta** [reply]
Thanks!

**Reply by MirazMac** [reply]
Any plans to include it in the Tauri core? It's cumbersome to worry about CORS in a "native" application. And Tauri's http plugin doesn't even work with thousands of existing popular libraries that use XHR or advanced fetch features like streaming.

**Reply by FabianLars** [reply]
Not at this point in time, maybe once more platforms have similar capabilities.

**Comment by 107295472** [comment]
> Thanks!
Why can't I find the cast?

**Reply by 107295472** [reply]
https://github.com/microsoft/windows-rs/blob/master/crates/libs/core/src/interface.rs

**Reply by FabianLars** [reply]
Hi 👋 You have to make sure to use the correct package versions, otherwise it won't work. As of Tauri 1.5 (and likely all following 1.x releases) this means webview2-com 0.19 and windows-rs 0.39.
Once the versions match the cast example in my code example above should work again.

**Reply by 107295472** [reply]
webview2-com = "0.26.0"
webview2-com-sys = "0.26.0"
windows = "0.51.1"
windows-sys = "0.48.0"

**Reply by FabianLars** [reply]
Like i said, those versions have to match what Tauri uses, so 0.19 for webview2-com and 0.39 for windows. If you use the versions you just posted it won't work.

**Reply by 107295472** [reply]
Thank you, I can change it to the version number you mentioned.

**Comment by cck190891** [comment]
Hello, I also have the same problem. Have you solved it?
this is my  tauri info 
```
[✔] Environment
    - OS: Windows 10.0.19045 X64
    ✔ WebView2: 122.0.2365.66
    ✔ MSVC: Visual Studio Build Tools 2022
    ✔ rustc: 1.75.0 (82e1608df 2023-12-21)
    ✔ cargo: 1.75.0 (1d8b05cdd 2023-11-20)
    ✔ rustup: 1.26.0 (5af9b9484 2023-04-05)
    ✔ Rust toolchain: stable-x86_64-pc-windows-msvc (default)
    - node: 20.10.0
    - pnpm: 8.15.1
    - npm: 10.2.3

[-] Packages
    - tauri [RUST]: 2.0.0-beta.8
    - tauri-build [RUST]: 2.0.0-beta.6
    - wry [RUST]: 0.37.0
    - tao [RUST]: 0.26.1
    - @tauri-apps/api [NPM]: 2.0.0-beta.3
    - @tauri-apps/cli [NPM]: 2.0.0-beta.6

[-] App
    - build-type: bundle
    - CSP: unset
    - frontendDist: ../dist
    - devUrl: http://localhost:1420/
    - framework: React
    - bundler: Vite
``` 
I've already tried this
```
webview2-com = "0.19.1"
windows = "0.39.0"
``` 
and run this code 
```
let core_2 = webview.controller().CoreWebView2().unwrap()
  .cast::<ICoreWebView2_2>()
  .unwrap();
```
but still get this 
```
cast method not found in `ICoreWebView2`
```
I think it's a version problem as mentioned above, but I don't know how to find the corresponding version

**Reply by 107295472** [reply]
You have to check which version 2.0 refers to

**Reply by cck190891** [reply]
I know I need to find the corresponding v2 version, but I'm not sure how to do it. Should I look for it through the Tauri version or the WebView version, or something else? Sorry, I'm a beginner in Rust, I have no idea about this.

**Reply by 107295472** [reply]
Go to his source code to find cargo.toml

**Reply by cck190891** [reply]
I found this at cargo.toml and tried this version but still get the same error
```
[target."cfg(windows)".dependencies.webview2-com]
version = "0.28"

[target."cfg(windows)".dependencies.windows]
features = ["Win32_Foundation"]
version = "0.52"
```

**Reply by 107295472** [reply]
Then I don’t know what the problem is, I’m using 1.x

**Reply by cck190891** [reply]
I've tried with Tauri 1.6.1 and its corresponding version, it works correctly. But I don't know why it's not working with version 2.0. Still, thank you for your assistance.

**Reply by FabianLars** [reply]
If you add `use windows::core::ComInterface;` to your rs file it should work. At least it does for me.

For reference in the versions used for v1 it was `use windows::core::Interface;`

If that still doesn't work, please cd into the src-tauri dir, run `cargo run` and show us the _full_ error message.

**Reply by cck190891** [reply]
Wow, it worked. 

I used ```use windows::core::Interface;``` before, but it always errored and told me I needed to use ```use windows_core::com_interface::ComInterface;```. I put it in, but it still didn't work for me. 

Now, using ```use windows::core::ComInterface;```, it works normally. 

Thank you for your help!

**Comment by MirazMac** [comment]
@FlysoftBeta did you find a solution for this? I'm looking for the same thing
